src/AddressInput/AddressInput.js

```diff
@@ -51,30 +48,102 @@ class AddressInput extends React.PureComponent {
     );
   };
 
-  render() {
+  _getInputValue = () => {
+    const { value: controlledValue } = this.props;
     const { inputValue } = this.state;
+
+    const value = controlledValue !== undefined ? controlledValue : inputValue;
+    return value;
+  };
+
+  _getIsLoading = () => {
+    const { status } = this.props;
+    return status === 'loading';
+  };
+
+  _setDropdownOpen = () => this.setState({ isDropdownOpen: true });
+
+  _setDropdownClosed = () => this.setState({ isDropdownOpen: false });
+
+  _renderLoadingOption = () => {
+    return {
+      id: '',
+      disabled: true,
+      value: () => (
+        <Box flex={1} align="center" padding="3px">
+          <Loader size="tiny" dataHook={dataHooks.loader} />
+        </Box>
+      ),
+    };
+  };
+
+  _renderNoResultsOption = () => {
+    const { noResultsText } = this.props;
+    const isString = typeof noResultsText === 'string';
+    const value = isString ? (
+      <Text light secondary dataHook={dataHooks.noResultsText}>
+        {noResultsText}
+      </Text>
+    ) : (
+      noResultsText
+    );
+
+    return {
+      id: '',
+      disabled: true,
+      overrideOptionStyle: !isString,
+      value,
+    };
+  };
+
+  _renderOptions = () => {
+    const { options } = this.props;
+    const value = this._getInputValue();
+    const isLoading = this._getIsLoading();
+    const noResultsFound = !options || !options.length;
```

Comment:
Just for readability:
```suggestion
    const noResultsFound = !options || options.length === 0;
```